---
title: 'Analyzing SARS-CoV-2 in Brazil Based on Patient Geo Location and Sex'
author: "Chaitrali Deshpande"
date: "`r format(Sys.time(), '%d %B, %Y')`"
output: pdf_document
csl: bioinformatics.csl
bibliography: references.bib
params:
  gff_file_path: "data/metadata/sars_refgenome_annotation.gff"
  vcf_dir_path: "data/11_vcf_output_for_R"
  sra_runtable_path: "data/00_sra_runtable/SraRunTable_Brazil.txt"
---

```{r setup, include = FALSE}
knitr::opts_chunk$set(echo = FALSE,
                      warning = FALSE,
                      message = FALSE,
                      error = TRUE)
```

# Background and Overview

The year 2020 was completely dominated by the COVID-19 pandemic.  It is still continuing to spread and mutate even after the discovery of numerous vaccines.  It has completely changed the way people interact with each other as the COVID-19 is an air-borne, contagious disease.  First discovered in Wuhan, China in December, 2019, COVID-19 is caused by severe acute respiratory syndrome coronovirus 2 (SARS-CoV-2).  Symptoms of COVID-19 include fever, chills, shortness of breath, exhaustion, headache, sore throat, nausea, and diarrhea.  The United States of America (US), India, and Brazil are the top three countries with the most COVID-19 cases and deaths caused by COVID-19.  The pandemic in Brazil is especially XXXXXXXXXXX given the smaller population of Brazil compared to US and India.

Between January 3rd and December 10th, 2021, there were 22,167,781 confirmed cases of COVID-19 in Brazil, with 616,251 deaths.  Brazil was averaging over 1000 deaths per day by June 2021.  The foreigners traveling from Europe to the different states of Brazil are considered the main reason for the widespread spread of the disease.  While he underdeveloped of north and northeast regions of Brazil are highly vulnerable, the metropolitan states Sao Paulo and Rio de Janeiro contributed to XXXXXX. 

I recently read an article on how Brazil is becoming the 'natural laboratory' for the COVID-19 virus and also a breeding ground for new variants of the virus. The Delta (B.1.617.2) variant cases were increased from 2.3 percent in June to 23.6 percent in July. It is a nine fold increase in the cases. The cases were emerging and it became a variant of concern. I was intrigued by the information and chose to work with this data.[@lamarca2021genomic] This variant of concern (VOC) is said to be more contagious than other variants and targets the non-vaccinated people.[@katella20215] The Delta virus mutations are concerning since it mutates numerous time to the spike protein unlike other variants.[@callaway2021mutation] It has 13 mutations that alters the amino-acid sequences of the protein it encodes.
I have worked on my project to analyze the impact of Delta Brazilian states and municipalities to find out more about how the they were affected by the Delta variants of SARS-CoV-2. This dataset includes timeframe from June, 2021 to July, 2021 when  the cases of Delta variant SARS-CoV-2 were on peak.

# Methods

## Genome Sequence Data 

I downloaded the Genome sequencing data of COVID-19 patients in municipalities of Brazil that were dominantly affected by COVID-19 states from the NCBI (National Center for Biotechnology Information) database.  It can be found on BioProject Resource database of NCBI titled 'SARS-CoV-2 genome sequencing Brazil' with Accession Number: PRJNA774631 and SRA Study: SRP343167.  It has total of 66 sequence data obtained using on Illumina platform from Brazil during June-July 2021.  

### Genome Sequence Processing Pipeline

* Download of the metadata:
Using SraRunTable I downloaded the metadata for Brazil from NCBI website. The downloaded raw sequences were pushed through 'fasterq-dump', a program written by NCBI to make the downloading of raw sequences feasible. The data was forked by
'fastqc' and analyzed for its quality.

* Eliminate low quality and unnecessary data:
'fastqc' tool is used to check the quality of raw data. The raw sequencing dataset contains a number of redundant data. To trim off the primers, delete low-quality base pairs and bad reads, 'trimmomatic' is used. It is an adapter to trim off primers and also gets rid of low quality bases and bad reads.

* Sorting the dataset:
I used bwa (Burrows-Wheeler Aligner) that created the index and maps the short reads against the reference genome.  Later the data goes to sam_and_bam files i.e. samtools and bamtools which sorted and processed the mapped reads.

* Data visualization and perl script:
To visualize the processed and mapped data, it is passed through perl scripting which takes mapped and processed sam_and_bam files and processes it into 'vcf' files.

* Graphical analysis:
I have used ggplots, ggthemes to create plots, geome, scales of my data to understand more about the sequences. The mapped and processed reads can be viewed in IGV

* Analyzing visuals and knitting:
Using vcfR the data is passed to 'R' for visualization and knitted using 'knitter'
to create the report of the analysis.

## Brazil COVID-19 Pandemic Statistics

In addition to the Genome data, I scoured the internet to obtain municipality-level statistical regarding the number of COVID-19 cases, deaths and vaccination rate in Brazil in the time period between June-July 2021.  The github repository [@CotaCovid19br2020] maintains temporal statistical data for COID-19 pandemic in Brazil.  I extracted the data relevant to this project i.e. data on 1st June, 2021 using excel and copied it in a csv file so that it can be imported by the R script.

## R Packages
tinytex: Using tinytex I have created pdf files for markdown.
ggplot2[@ggplot2]: Using ggplot2 I have created graphics for my data analysis.
ggthemes[@ggthemes]: Downloaded ggthemes to get an additional gnome, scales and ggplot for my analysis.
hrbrthemes[@hrbrthmes]and viridis[@viridis]: Using hrbrthemes to color and plotted labels to the graphs.
knitter: Using 'knitter' a report is created including analysis. 
Rcpp[@eddelbuettel2011rcpp]: The Rcpp package provided a consistent API for seamlessly accessing, extending or modifying R objects at the C++ level.
janitor[@janitor]: To clean unnecessary data frames, I have used janitor package for my analysis.
deplore and tidyr[@tidyr]: To categorize and parse the data I have used dplyr and tidyr.
readr[@readr]: I have used readR to read in and write the csv files of my data folder.
magrittr and stringr[@stringr][@magrittr]: To make data tables and piping them as well as to avoid nested function calls and minimize the need for local variables and definition in my analysis, I have used magrittr. 

# Results and Discussion

## Brazil Sequencing Data

### People Infected with SARS-CoV-2 from Different Brazilian municipalities of Rio de Janerio:

After analyzing the data from June, 2020 to July, 2020 it is observed that the given data downloaded from NCBI  is correct when verified with other sources on the internet. The Brazilian state Rio De Janerio has underdeveloped and vulnerable municipality population. 

## Mutations in the variant:

After analyzing the data of SNP's, the count of S genes are highest as compared to other followed by N. The data is collected from June and July 2020 when the variant was continuously mutating. It is concluded by the graph that Gene S has highest length 3821 followed by Gene N which explains the rate of mutation in them will be faster as they undergo the transcription and translation processes. 

Total count of mutations is higher in female when compared to male. It can be concluded that the population of female in the cities with greater number of Covid-19 cases is higher than male. That means more women in those area were affected. However, I am unsure if the SNP's have any kind of relation with gender.  

## Mutations of the variants: 

The number of death in these municipalities during the time shows the effects of mutations of the variant and the analysis is correct. The Delta variant seems to have widely impacted the cases during that time. The cases of mutations are higher in Rio de Janerio, Mesquita and Duque de Caxias cities for male populatiion. And the rate of mutation is higher in Mesquita, Rio de Janerio and Nova Iguacu for female population. I concluded that the overall female population was infected with the Delta variant. I am not sure if it related to gender. 

* 1. Count of distinct SNPs in Named SARS-CoV-2 Genes:
I plotted the SNPs (single nucleotide polymorphism) against named SARS-CoV-2. 

* 2. Table containing gene name and length;
From the data in the table it is concluded that Gene S has highest length 3821 and ORF10
has the lowest 116. The ORF7a and ORF8 have same length i.e. 365.

*3. Analysis of samples from male and female from different municipalities:
The data in the table shows the total number of samples analysed for the SARS-CoV-2 
mutation. The samples from five municipalities were disregarded since the number of 
samples given were zero. 

* 4. Count of distinct SNP's in male and female:
The graph shows the information how SNP's have increased in the set of male sample analyzed.
The data was calculated by dividing the count of genes by number of sample analyzed.

* 5.  
Total mutations divided by the number of sample:
This graph shows the total mutations count is somewhat similar in male and female for Japeri municipality.
The reason could be the  

# Conclusion

```{r load-packages-and-functions}
library("vcfR")
library("ggplot2")
library("dplyr")
library("ggthemes")
library("readr")
library("magrittr")
library("kableExtra")
library("tidyr")
library("gridExtra")

# this loads all function files in the code/functions directory
function_paths <- list.files("code/functions", full.names = TRUE)
purrr::walk(function_paths, source)
```

```{r make-directories}
# create necessary directories to organize output
dir.create(file.path("data", "processed_data"), showWarnings = FALSE)
dir.create(file.path("output", "figures"), showWarnings = FALSE)
dir.create(file.path("output", "tables"), showWarnings = FALSE)
```

```{r load-vcf-data}
# load in, tidy, and stack all of the VCF files in the target directory
# these are the output of the bash pipeline
stacked_vcfs <- parse_tidy_and_stack_vcfs(
  vcf_dir_path = params$vcf_dir_path)

# write Stacked VCF data to a csv file to processed data folder
write_csv(stacked_vcfs, file.path("data", "processed_data",
                                         "stacked_vcfs.csv"))

# load in the gff file with genome annotations for SARS-CoV-2
gff <- read_gff(gff_file_path = params$gff_file_path)

# pull out just the named genes from the annotation file
gene_table <- extract_genes_from_gff(annotation_object = gff)

# combine the stacked vcf table with the gene names and the metadata from
# the sra runtable file
vcf_with_metadata <- add_genes_metadata_to_vcfstack(
  sra_runtable_path = params$sra_runtable_path,
  stacked_vcf = stacked_vcfs,
  cleaned_genes_table = gene_table)

#Clean geo location name
vcf_with_metadata <- vcf_with_metadata %>%
  tidyr::separate(geo_loc_name, c(NA, "Municipalities"), sep = ",")

# write combined stacked vcf and metadata to output folder
write_csv(vcf_with_metadata, file.path("data", "processed_data",
                                              "vcf_with_metadata.csv"))
```

```{r load-sra-data}
# read metadata
sraruntable <- read_csv(params$sra_runtable_path)
sraruntable <- sraruntable %>%
  tidyr::separate(geo_loc_name, c(NA, "Municipalities"), sep = ",")

# count Municipalities wise female samples in the metadata
female_table <- sraruntable %>%
  filter(host_sex == "female") %>%
  count(Municipalities) %>%
  rename(female_count = n)

# count Municipalities wise female samples in the metadata
male_table <- sraruntable %>%
  filter(host_sex == "male") %>%
  count(Municipalities) %>%
  rename(male_count = n)

# count Municipalities wise total samples in the metadata
total_table <- sraruntable %>%
  count(Municipalities, sort = TRUE) %>%
  rename(total_count = n)

# merge tables and remove NA
final_table <- left_join(left_join(total_table, female_table), male_table)
final_table[is.na(final_table)] <- 0
``` 

```{r load-brazil-data}
brazildata <- read_csv(file.path("data", "raw_data",
                                  "brazil_covid_data_citywise.csv"))

tidybrazildata <- brazildata[, c("date", "city",
                                "totalCases_per_100k_inhabitants",
                                "vaccination_rate")]
```

# Figures

```{r figure1}
# create a plot of unique SNP locations within each gene across all samples
figure1 <- vcf_with_metadata %>%
  filter(!is.na(gene)) %>% # get rid of SNPs not in gene regions
  group_by(gene, pos) %>%
  tally() %>% # this gives a column n for the number of gene by position
  group_by(gene) %>%
  tally() %>% # this collapses that down to the number of unique SNP locations
  ggplot(aes(x = gene,
             y = n)) +
    geom_col() +
    labs(title = "Count of distinct SNPs in Named SARS-CoV-2 Genes",
         x = "Gene Name", y = "Count") +
  theme_few() # get rid of the grey background

  ggsave(figure1, filename = paste0("output/figures/figure1.png"))
  figure1
```

**Figure 1**: N and S genes have more unique SNPs in the set of samples analyzed.

```{r figure2}
# create a plot of unique SNP locations within each gene across different sex
figure2 <- vcf_with_metadata %>%
  filter(!is.na(gene)) %>% # get rid of SNPs not in gene regions
  group_by(gene, host_sex, pos) %>%
  tally() %>% # this gives a column n for the number of gene by position
  group_by(gene, host_sex) %>%
  tally() %>% # this collapses that down to the number of unique SNP locations
  ggplot(aes(x = gene, y = n, fill = host_sex)) +
    geom_bar(stat = "identity", position = "dodge") + # makes a bar plot
    labs(title = "Count of distinct SNPs in Named SARS-CoV-2 Genes",
         x = "Gene Name", y = "Count", fill = "Host sex") +
            #add title, axis labels, legend label
    geom_col(aes(fill = host_sex), position = "dodge") +
    geom_text(aes(label = n), position = position_dodge(0.9), vjust = 1.5) +
        # add labels to the bars
    theme_few() # get rid of the grey background
ggsave(figure2, filename = paste0("output/figures/figure2.png"))
figure2
```

**Figure 2**: N and S genes have more unique SNPs in the set of samples analyzed.

```{r figure3}
# create a plot of unique SNP locations within each gene across different sex
figure3 <- vcf_with_metadata %>%
  group_by(Municipalities, host_sex) %>%
  filter(!Municipalities %in% c("ITABORAI", "ITAGUAI", "NITEROI",
                                "NOVA BASSANO", "SANTO ANTONIO DE PADUA",
                                "XAMBIOA")) %>%
  tally() %>%   # this gives a column n for the number of gene by position
  ggplot(aes(x = Municipalities, y = n, fill = host_sex)) +
      geom_bar(stat = "identity", position = "dodge") + # makes a bar plot
      labs(title = "Total Mutations", x = "Municipality", y = "Count",
         fill = "Host Sex") + #add title, axis labels, legend label
      geom_col(aes(fill = host_sex), position = "dodge") +
      geom_text(aes(label = n), position = position_dodge(0.9), vjust = 1.5) +
        # add labels to the bars
      theme_few() +  # get rid of the grey background
      theme(axis.text.x = element_text(angle = 90, vjust = 0.5, hjust = 1))
        # makes the x-label vertical
ggsave(figure3, filename = paste0("output/figures/figure3.png"))
figure3
```

**Figure 3**: N and S genes have more unique SNPs in the set of samples analyzed.


```{r figure4}
# create a plot of unique SNP locations within each gene across different sex
total_mutations_sex <- vcf_with_metadata %>%
  group_by(Municipalities, host_sex) %>%
  filter(!Municipalities %in% c("ITABORAI", "ITAGUAI", "NITEROI",
                                "NOVA BASSANO", "SANTO ANTONIO DE PADUA",
                                "XAMBIOA")) %>%
  tally() %>%
  rename(mutation_count = n)
  
total_mutations_sex <- rbind(left_join(total_mutations_sex %>%
                              filter(host_sex == "male"), male_table),
                             left_join(total_mutations_sex %>%
                              filter(host_sex == "female"), female_table)) %>%
                  mutate(sample_count = coalesce(male_count, female_count)) %>%
                  mutate(normalized = round(mutation_count / sample_count, 0))

# this gives a column n for the number of gene by position
figure4 <- ggplot(total_mutations_sex,
                  aes(x = Municipalities, y = normalized, fill = host_sex)) +
  geom_bar(stat = "identity", position = "dodge") + # makes a bar plot
  labs(title = "Total Mutations Divided by Number of Samples",
       x = "Municipality", y = "Count", fill = "Host Sex") +
          #adds title, axis labels, legend label
  geom_col(aes(fill = host_sex), width = 0.5, position = position_dodge(0.5)) +
  geom_text(aes(label = normalized),
            position = position_dodge(0.9),
            vjust = 1.5) +
      # add labels to the bars
  theme_few() +  # get rid of the grey background
  theme(axis.text.x = element_text(angle = 90, vjust = 0.5, hjust = 1))
      # makes the x-label vertical
ggsave(figure4, filename = paste0("output/figures/figure4.png"))
figure4
```

**Figure 4**: N and S genes have more unique SNPs in the set of samples analyzed.

```{r figure5}
# create a plot of unique SNP locations within each gene across different sex
figure5 <- vcf_with_metadata %>%
  group_by(Municipalities, host_sex, pos) %>%
  filter(!Municipalities %in% c("ITABORAI", "ITAGUAI", "NITEROI",
                                "NOVA BASSANO", "SANTO ANTONIO DE PADUA",
                                "XAMBIOA")) %>%
  tally() %>%
  group_by(Municipalities, host_sex) %>%
  tally() %>%
  # this gives a column n for the number of gene by position
  ggplot(aes(x = Municipalities, y = n, fill = host_sex)) +
      geom_bar(stat = "identity", position = "dodge") + # makes a bar plot
      labs(title = "Total Unique SNPs", x = "Municipality", y = "Count",
           fill = "Host Sex") + #add title, axis labels, legend label
      geom_col(aes(fill = host_sex), position = "dodge") +
      geom_text(aes(label = n), position = position_dodge(0.9), vjust = 1.5) +
      # add labels to the bars
      theme_few() +  # get rid of the grey background
      theme(axis.text.x = element_text(angle = 90, vjust = 0.5, hjust = 1))
        # makes the x-label vertical
ggsave(figure5, filename = paste0("output/figures/figure5.png"))
figure5
```

**Figure 5**: N and S genes have more unique SNPs in the set of samples analyzed.

```{r figure6}
# create a plot of unique SNP locations within each gene across different sex

# total_mutations <- vcf_with_metadata %>%
#   group_by(Municipalities) %>%
#   filter(!Municipalities %in% c("ITABORAI", "ITAGUAI", "NITEROI",
#                                 "NOVA BASSANO", "SANTO ANTONIO DE PADUA",
#                                 "XAMBIOA")) %>%
#   tally()
#
# total_mutations <- left_join(total_mutations, final_table) %>%
#                    mutate(normalized = round(n / total_count, 0))

total_snp <- vcf_with_metadata %>%
  group_by(Municipalities, pos) %>%
  filter(!Municipalities %in% c("ITABORAI", "ITAGUAI", "NITEROI",
                                "NOVA BASSANO", "SANTO ANTONIO DE PADUA",
                                "XAMBIOA")) %>%
  tally() %>%
  group_by(Municipalities) %>%
  tally()

figure6 <- left_join(total_snp, tidybrazildata,
                     by = c("Municipalities" = "city")) %>%
  ggplot(aes(x = n, y = totalCases_per_100k_inhabitants)) +
  geom_point(aes(color = Municipalities)) +
  labs(title = "Total Cases vs. Total Unique SNPs",
       x = "Count of unique SNPs", y = "Total cases per 100k inhabitants") +
      #add title, axis labels, legend label
    # geom_text(aes(label = n), position = position_dodge(0.9), hjust=1.5) +
    # # add labels to the bars
  theme_few() # get rid of the grey background
ggsave(figure6, filename = paste0("output/figures/figure6.png"))
figure6
```

**Figure 6**: N and S genes have more unique SNPs in the set of samples analyzed.

```{r figure7}
figure7 <- left_join(total_snp, tidybrazildata,
                     by = c("Municipalities" = "city")) %>%
  ggplot(aes(x = n, y = vaccination_rate)) +
  geom_point(aes(color = Municipalities)) +
  labs(title = "Total Cases vs. Total Unique SNPs",
       x = "Count of unique SNPs", y = "Total cases per 100k inhabitants") +
    #add title, axis labels, legend label
  # geom_text(aes(label = n), position = position_dodge(0.9), hjust=1.5) +
  # # add labels to the bars
  theme_few() # get rid of the grey background

# g_legend<-function(a.gplot) {
#   tmp <- ggplot_gtable(ggplot_build(a.gplot))
#   leg <- which(sapply(tmp$grobs, function(x) x$name) == "guide-box")
#   legend <- tmp$grobs[[leg]]
#   return(legend)}
#
# mylegend<-g_legend(p1)
#
# p3 <- grid.arrange(arrangeGrob(p1 + theme(legend.position="none"),
#                          p2 + theme(legend.position="none"),
#                          nrow=1),
#              mylegend, nrow=2,heights=c(10, 1))
#
# ggsave(p3, filename = paste0("output/figures/figure6.png"))

ggsave(figure7, filename = paste0("output/figures/figure7.png"))
figure7
```

**Figure 7**: N and S genes have more unique SNPs in the set of samples analyzed.

\vspace{5mm}

# Tables

```{r table1}
# An example table to show the length of each gene using its start and end
gene_table %>%
  mutate(length = end - start) %>%
  select(gene_name, start, end, length) %>%
  knitr::kable(col.names = c("Gene Name", "Sequence Start",
                             "Sequence End", "Gene Length"),
                              align = "c")
```

**Table 1**: Gene names, locations, and lengths in the SARS-CoV-2 genome. Higher SNP counts in the S and N genes may be related to the larger size of these genes.

\vspace{5mm}

```{r table2}
final_table %>%  knitr::kable(col.names =
                                   c("Municipality",
                                     "Total sample count",
                                     "Female Count", "Male Count"),
                              align = "c")
```

**Table 2**: Gene names, locations, and lengths in the SARS-CoV-2 genome. Higher SNP counts in the S and N genes may be related to the larger size of these genes.

\vspace{5mm}

```{r table3}
# An example table to show the length of each gene using its start and end
tidybrazildata %>%  knitr::kable(col.names =
                                   c("Date", "Municipality",
                                     "Total Cases per 100k Inhabitants",
                                     "Fully Vaccination Rate (%)"),
                                 align = "c")
```

**Table 3**: Gene names, locations, and lengths in the SARS-CoV-2 genome. Higher SNP counts in the S and N genes may be related to the larger size of these genes.

# Sources Cited
