---
title: 'Analyzing SARS-CoV-2 in Brazil Based on Patient Geo Location and Sex'
author: "Chaitrali Deshpande"
date: "`r format(Sys.time(), '%d %B, %Y')`"
output: html_document
csl: bioinformatics.csl
bibliography: references.bib
params:
  gff_file_path: "data/metadata/sars_refgenome_annotation.gff"
  vcf_dir_path: "data/11_vcf_output_for_R"
  sra_runtable_path: "data/00_sra_runtable/SraRunTable_Brazil.txt"
---

```{r setup, include = FALSE}
knitr::opts_chunk$set(echo = FALSE,
                      warning = FALSE,
                      message = FALSE,
                      error = TRUE)
```

# Background and Overview

This is a report on SARS-CoV-2, including some variant analysis [@koyama2020variant].

# Methods

See the set of tutorials on the [vcfR package website](https://knausb.github.io/vcfR_documentation/index.html).

You may also want to use any of a range of different COVID data packages and data sources:

* https://kjhealy.github.io/covdata/
* https://github.com/como-ph/oxcovid19
* https://ropensci.org/blog/2020/10/20/searching-medrxivr-and-biorxiv-preprint-data/
* https://covidtracking.com/data/api
  * `readr::read_csv("https://api.covidtracking.com/v1/states/daily.csv")`

## Subsections are ok too

# Results and Discussion

```{r load-packages-and-functions}
library("vcfR")
library("ggplot2")
library("dplyr")
library("ggthemes")
library("readr")
library("magrittr")
library("kableExtra")
library("tidyr")
library("gridExtra")

# this loads all function files in the code/functions directory
function_paths <- list.files("code/functions", full.names = TRUE)
purrr::walk(function_paths, source)
```

```{r make directories}
# create necessary directories to organize output
dir.create(file.path("data", "processed_data"), showWarnings = FALSE)
dir.create(file.path("output", "figures"), showWarnings = FALSE)
dir.create(file.path("output", "tables"), showWarnings = FALSE)
```

```{r load-vcf-data}
# load in, tidy, and stack all of the VCF files in the target directory
# these are the output of the bash pipeline
stacked_vcfs <- parse_tidy_and_stack_vcfs(
  vcf_dir_path = params$vcf_dir_path)

# write Stacked VCF data to a csv file to processed data folder
write_csv(stacked_vcfs, file.path("data", "processed_data", 
                                         "stacked_vcfs.csv"))

# load in the gff file with genome annotations for SARS-CoV-2
gff <- read_gff(gff_file_path = params$gff_file_path)

# pull out just the named genes from the annotation file
gene_table <- extract_genes_from_gff(annotation_object = gff)

# combine the stacked vcf table with the gene names and the metadata from
# the sra runtable file
vcf_with_metadata <- add_genes_metadata_to_vcfstack(
  sra_runtable_path = params$sra_runtable_path,
  stacked_vcf = stacked_vcfs,
  cleaned_genes_table = gene_table)

#Clean geo location name
vcf_with_metadata <- vcf_with_metadata %>%
  tidyr::separate(geo_loc_name, c(NA,"Municipalities"), sep = ",")

# write combined stacked vcf and metadata to output folder
write_csv(vcf_with_metadata, file.path("data", "processed_data",
                                              "vcf_with_metadata.csv"))
```

```{r load-sra-data}
# read metadata
sraruntable <- read_csv(params$sra_runtable_path)
sraruntable <- sraruntable %>%
  tidyr::separate(geo_loc_name, c(NA,"Municipalities"), sep = ",")

# count Municipalities wise female samples in the metadata
female_table <- sraruntable %>% filter(host_sex == "female") %>% 
  count(Municipalities) %>%
  rename(Female = n)

# count Municipalities wise female samples in the metadata
male_table <- sraruntable %>% filter(host_sex == "male") %>% 
  count(Municipalities) %>%
  rename(Male = n)

# count Municipalities wise total samples in the metadata
total_table <- sraruntable %>% count(Municipalities, sort = TRUE) %>%
  rename(Total = n)

# merge tables and remove NA
final_table <- left_join(left_join(total_table, female_table), male_table)
final_table[is.na(final_table)] = 0
``` 

```{r load-brazil-data}
brazildata <- read_csv(file.path("data", "raw_data",
                                  "brazil_covid_data_citywise.csv"))

tidybrazildata <- brazildata[,c("date", "city", "totalCases_per_100k_inhabitants", "vaccination_rate")]
```

# Figures

```{r figure1}
# create a plot of unique SNP locations within each gene across all samples
figure1 <- vcf_with_metadata %>%
  filter(!is.na(gene)) %>% # get rid of SNPs not in gene regions
  group_by(gene, pos) %>%
  tally() %>% # this gives a column n for the number of gene by position
  group_by(gene) %>%
  tally() %>% # this collapses that down to the number of unique SNP locations
  ggplot(aes(x = gene,
             y = n)) +
    geom_col() +
    labs(title = "Count of distinct SNPs in Named SARS-CoV-2 Genes",
         x = "Gene Name", y = "Count") +
  theme_few() # get rid of the grey background

  ggsave(figure1, filename = paste0("output/figures/figure1.png"))
  figure1
```

**Figure 1**: N and S genes have more unique SNPs in the set of samples analyzed.

```{r figure2}
# create a plot of unique SNP locations within each gene across different sex
figure2 <- vcf_with_metadata %>%
  filter(!is.na(gene)) %>% # get rid of SNPs not in gene regions
  group_by(gene, host_sex, pos) %>%
  tally() %>% # this gives a column n for the number of gene by position
  group_by(gene, host_sex) %>% 
  tally() %>% # this collapses that down to the number of unique SNP locations
  ggplot(aes(x = gene, y = n, fill = host_sex)) +
    geom_bar(stat='identity', position='dodge') + # makes a bar plot
    labs(title = "Count of distinct SNPs in Named SARS-CoV-2 Genes",
         x = "Gene Name", y = "Count", fill = "Host sex") + #add title, axis labels, legend label
    geom_col(aes(fill = host_sex), position = "dodge") +
    geom_text(aes(label = n), position = position_dodge(0.9), vjust=1.5) +
        # add labels to the bars
    theme_few() # get rid of the grey background
  ggsave(figure2, filename = paste0("output/figures/figure2.png"))
  figure2
```

**Figure 2**: N and S genes have more unique SNPs in the set of samples analyzed.

```{r figure3}
# create a plot of unique SNP locations within each gene across different sex
figure3 <- vcf_with_metadata %>%
  group_by(Municipalities, host_sex) %>%
  filter(!Municipalities %in% c("ITABORAI", "ITAGUAI", "NITEROI", "NOVA BASSANO",
                               "SANTO ANTONIO DE PADUA", "XAMBIOA")) %>%
  tally() %>%   # this gives a column n for the number of gene by position
  ggplot(aes(x = Municipalities, y = n, fill = host_sex)) +
      geom_bar(stat='identity', position='dodge') + # makes a bar plot
      labs(title = "Total Mutations", x = "Municipality", y = "Count", 
         fill = "Host Sex") + #add title, axis labels, legend label
      geom_col(aes(fill = host_sex), position = "dodge") +
      geom_text(aes(label = n), position = position_dodge(0.9), vjust=1.5) +
        # add labels to the bars
      theme_few() +  # get rid of the grey background
      theme(axis.text.x = element_text(angle = 90, vjust = 0.5, hjust=1))
        # makes the x-label vertical
  ggsave(figur3, filename = paste0("output/figures/figure3.png"))
  figure3
```

**Figure 3**: N and S genes have more unique SNPs in the set of samples analyzed.


```{r figure4}
# create a plot of unique SNP locations within each gene across different sex
total_mutations_sex <- vcf_with_metadata %>%
  group_by(Municipalities, host_sex) %>%
  filter(!Municipalities %in% c("ITABORAI", "ITAGUAI", "NITEROI", "NOVA BASSANO",
                               "SANTO ANTONIO DE PADUA", "XAMBIOA")) %>%
  tally()
  
total_mutations_sex <- rbind(left_join(total_mutations_sex %>% filter(host_sex=="male"), male_table),
      left_join(total_mutations_sex %>% filter(host_sex=="female"), female_table)) %>%
  mutate(normalized = round(n / sample_count,0))

# this gives a column n for the number of gene by position
figure4 <- ggplot(total_mutations_sex, aes(x = Municipalities, y = normalized, fill = host_sex)) +
    geom_bar(stat='identity', position='dodge') + # makes a bar plot
    labs(title = "Total Mutations Divided by Number of Samples", x = "Municipality", y = "Count", 
       fill = "Host Sex") + #add title, axis labels, legend label
    geom_col(aes(fill = host_sex), width = 0.5, position = position_dodge(0.5)) +
    geom_text(aes(label = normalized), position = position_dodge(0.9), vjust=1.5) +
      # add labels to the bars
    theme_few() +  # get rid of the grey background
    theme(axis.text.x = element_text(angle = 90, vjust = 0.5, hjust=1))
      # makes the x-label vertical
  ggsave(figure4, filename = paste0("output/figures/figure4.png"))
  figure4
```

**Figure 4**: N and S genes have more unique SNPs in the set of samples analyzed.

```{r figure5}
# create a plot of unique SNP locations within each gene across different sex
figure5 <- vcf_with_metadata %>%
  group_by(Municipalities, host_sex, pos) %>%
  filter(!Municipalities %in% c("ITABORAI", "ITAGUAI", "NITEROI", "NOVA BASSANO",
                               "SANTO ANTONIO DE PADUA", "XAMBIOA")) %>%
  tally() %>%
  group_by(Municipalities, host_sex) %>%
  tally() %>%
  # this gives a column n for the number of gene by position
  ggplot(aes(x = Municipalities, y = n, fill = host_sex)) +
      geom_bar(stat='identity', position='dodge') + # makes a bar plot
      labs(title = "Total Unique SNPs", x = "Municipality", y = "Count",
           fill = "Host Sex") + #add title, axis labels, legend label
      geom_col(aes(fill = host_sex), position = "dodge") +
      geom_text(aes(label = n), position = position_dodge(0.9), vjust=1.5) +
      # add labels to the bars
      theme_few() +  # get rid of the grey background
      theme(axis.text.x = element_text(angle = 90, vjust = 0.5, hjust=1))
        # makes the x-label vertical
  ggsave(figure5, filename = paste0("output/figures/figure5.png"))
  figure5
```

**Figure 5**: N and S genes have more unique SNPs in the set of samples analyzed.

```{r figure6}
# create a plot of unique SNP locations within each gene across different sex

total_mutations <- vcf_with_metadata %>%
  group_by(Municipalities) %>%
  filter(!Municipalities %in% c("ITABORAI", "ITAGUAI", "NITEROI", "NOVA BASSANO",
                               "SANTO ANTONIO DE PADUA", "XAMBIOA")) %>%
  tally()

total_mutations <- left_join(total_mutations, final_table) %>%
                          mutate(normalized = round(n / Total, 0))

total_SNP <- vcf_with_metadata %>%
  group_by(Municipalities, pos) %>%
  filter(!Municipalities %in% c("ITABORAI", "ITAGUAI", "NITEROI", "NOVA BASSANO",
                               "SANTO ANTONIO DE PADUA", "XAMBIOA")) %>%
  tally() %>%
  group_by(Municipalities) %>%
  tally()


p1 <- left_join(total_SNP, tidybrazildata, by = c("Municipalities" = "city") ) %>% 
    ggplot(aes(x = n, y = totalCases_per_100k_inhabitants)) +
    geom_point(aes(color = Municipalities)) +
    labs(title = "Total Cases vs. Total Unique SNPs", 
         x = "Count of unique SNPs", y = "Total cases per 100k inhabitants") + 
      #add title, axis labels, legend label
    # geom_text(aes(label = n), position = position_dodge(0.9), hjust=1.5) +
    # # add labels to the bars
    theme_few() + theme(legend.position="bottom") # get rid of the grey background

p2 <- left_join(total_SNP, tidybrazildata, by = c("Municipalities" = "city") ) %>% 
    ggplot(aes(x = n, y = vaccination_rate)) +
    geom_point(aes(color = Municipalities)) +
    labs(title = "Total Cases vs. Total Unique SNPs", 
         x = "Count of unique SNPs", y = "Total cases per 100k inhabitants") + 
      #add title, axis labels, legend label
    # geom_text(aes(label = n), position = position_dodge(0.9), hjust=1.5) +
    # # add labels to the bars
    theme_few() # get rid of the grey background

g_legend<-function(a.gplot){
  tmp <- ggplot_gtable(ggplot_build(a.gplot))
  leg <- which(sapply(tmp$grobs, function(x) x$name) == "guide-box")
  legend <- tmp$grobs[[leg]]
  return(legend)}

mylegend<-g_legend(p1)

p3 <- grid.arrange(arrangeGrob(p1 + theme(legend.position="none"),
                         p2 + theme(legend.position="none"),
                         nrow=1),
             mylegend, nrow=2,heights=c(10, 1))

ggsave(p3, filename = paste0("output/figures/figure6.png"))
```

**Figure 6**: N and S genes have more unique SNPs in the set of samples analyzed.

# Tables

```{r table1}
# An example table to show the length of each gene using its start and end

library(magrittr)
library(kableExtra)

gene_table %>%
  mutate(length = end - start) %>%
  select(gene_name, start, end, length) %>%
  knitr::kable(col.names = c("Gene Name", "Sequence Start" , "Sequence End" , "Gene Length"),
               align = 'c', 
               caption = "Table 1: An example table caption it is long.", booktabs = TRUE) %>% 
    kable_styling(bootstrap_options = "striped")
```

**Table 1**: Gene names, locations, and lengths in the SARS-CoV-2 genome. Higher SNP counts in the S and N genes may be related to the larger size of these genes.

```{r table2}
final_table %>%  knitr::kable(col.names = 
                                   c("Municipality",
                                     "Total sample count",
                                     "Female Count", "Male Count"), align = 'c') %>% 
    kable_styling(bootstrap_options = "striped")
```
**Table 2**: Gene names, locations, and lengths in the SARS-CoV-2 genome. Higher SNP counts in the S and N genes may be related to the larger size of these genes.

```{r table3}
# An example table to show the length of each gene using its start and end
tidybrazildata %>%  knitr::kable(col.names = 
                                   c("Date", "Municipality",
                                     "Total Cases per 100k Inhabitants",
                                     "Fully Vaccination Rate (%)"), align = 'c') %>% 
    kable_styling(bootstrap_options = "striped")
```

**Table 3**: Gene names, locations, and lengths in the SARS-CoV-2 genome. Higher SNP counts in the S and N genes may be related to the larger size of these genes.

# Sources Cited
